kore/src/main/scala/org/kframework/attributes/Att.scala

```diff
@@ -40,130 +62,245 @@ class Att private (val att: Map[(String, String), Any]) extends AttributesToStri
+  final val PRIORITIES = Key("priorities")
+  final val PRIORITY = Key("priority")
+  final val PRIVATE = Key("private")
+  final val PRODUCTION = Key("org.kframework.definition.Production")
+  final val PRODUCTION_ID = Key("productionId")
+  final val PROJECTION = Key("projection")
+  final val PUBLIC = Key("public")
+  final val RECORD_PRD = Key("recordPrd")
+  final val RECORD_PRD_ZERO = Key("recordPrd-zero")
+  final val RECORD_PRD_ONE = Key("recordPrd-one")
+  final val RECORD_PRD_MAIN = Key("recordPrd-main")
+  final val RECORD_PRD_EMPTY = Key("recordPrd-empty")
+  final val RECORD_PRD_SUBSORT = Key("recordPrd-subsort")
+  final val RECORD_PRD_REPEAT = Key("recordPrd-repeat")
+  final val RECORD_PRD_ITEM = Key("recordPrd-item")
+  final val REFRESHED = Key("refreshed")
+  final val REGEX = Key("regex")
+  final val REJECT = Key("reject")
+  final val REMOVE = Key("remove")
+  final val RESULT = Key("result")
+  final val RETURNS_UNIT = Key("returnsUnit")
+  final val RIGHT = Key("right")
+  final val SEQSTRICT = Key("seqstrict")
+  final val SIMPLIFICATION = Key("simplification")
+  final val SMTLIB = Key("smtlib")
+  final val SMT_HOOK = Key("smt-hook")
+  final val SMT_LEMMA = Key("smt-lemma")
+  final val SMT_PRELUDE = Key("smt-prelude")
+  final val SORT = Key("org.kframework.kore.Sort")
+  final val SORT_PARAMS = Key("sortParams")
+  final val SOURCE = Key("org.kframework.attributes.Source")
+  final val STREAM = Key("stream")
+  final val STRICT = Key("strict")
+  final val SYMBOL = Key("symbol")
+  final val SYMBOLIC = Key("symbolic")
+  final val SYNTAX_MODULE = Key("syntaxModule")
+  final val TAG = Key("tag")
+  final val TEMPORARY_CELL_SORT_DECL = Key("temporary-cell-sort-decl")
+  final val TERMINALS = Key("terminals")
+  final val TOKEN = Key("token")
+  final val TOPCELL = Key("topcell")
+  final val TOTAL = Key("total")
+  final val TRANSITION = Key("transition")
+  final val TRUSTED = Key("trusted")
+  final val TYPE = Key("type")
+  final val UNBLOCK = Key("unblock")
+  final val UNBOUND_VARIABLES = Key("unboundVariables")
+  final val UNIT = Key("unit")
+  final val UNIQUE_ID = Key("UNIQUE_ID")
+  final val UNPARSE_AVOID = Key("unparseAvoid")
+  final val UNUSED = Key("unused")
+  final val USER_LIST = Key("userList")
+  final val USER_LIST_TERMINATOR = Key("userListTerminator")
+  final val WITH_CONFIG = Key("withConfig")
+  final val WRAP_ELEMENT = Key("wrapElement")
 
   private val stringClassName = classOf[String].getName
   private val intClassName = classOf[java.lang.Integer].getName
 
-  def from(thatAtt: java.util.Map[String, String]): Att =
+  // Marker for group(...) attributes
+  case class GroupMarker() extends AttValue
+  private val groupMarkerClassName = classOf[GroupMarker].getName
+
+  // All String fields with UPPER_CASE naming are considered part of the whitelist
+  private val whitelist: Set[Key] =
+    Att.getClass.getDeclaredFields
```

Comment:
I agree, however, one thing @radumereuta pointed out is that we want to ensure we can still dump code after some execution then reload it, so we don't want to outright ban these internal attributes from appearing in the source.

Perhaps we could add an `internal(_)` attribute that is sort of analogous to how `group(_)` works?

Even if we error check these internal attributes, it's also just a very leaky abstraction to have user and internal attributes in the same namespace. Every attribute we add internally to the compiler then becomes a reserved word in the language semantics and needs to be documented. 

In the frontend, we could easily fix all these issues by creating three different key classes:
- `Att.BuiltinKey`, for documented built-ins that can appear in user code
- `Att.InternalKey`, for compiler-internal attributes that can only appear in source code via `internal(my-internal-att)`
- `Att.UserGroup`, for user-defined groups that are defined via `group(my-group)`

We then give `Att.UserGroup` and `Att.InternalKey` separate namespaces, but still ensure that neither conflicts with any `Att.BuiltinKey`.

However, when emitting to KORE all the attributes are flattened, so this would take a bit more work to handle across the backends. Perhaps when emitting we could prefix builtin and internal attributes with some char that is not permitted to appear in user-defined groups?

Reply:
You raise some good points, but also some things I'm less sure of. For example, it's probably not going to be practical in general to kompile a definition, emit it as k after being compiled, and then recompile it and end up with the same compiled definition you got the first time. You can't do that now and I doubt you'll be able to in the future. So we don't need to worry about making sure that that can work.

Similarly, there is no reason to emit partially compiled definitions, mess with them manually, and then pass them back to the compiler to compile the rest of the way. This is a huge nightmare from an abstraction perspective because it causes the entire internal state of the compiler at every step to become part of the compiler's public API, which we definitely don't want. I would not try to make sure this is possible either, as it greatly complexifies the way tooling gets written.

As such, I don't really see the benefit of being able to parse internal attributes added by the compilation process. We do want to be able to parse attributes added by the /parsing/ process, since we probably want to be able to parse a definition, compile it, and then use the compiled and parsed definitions to modify the parsed definition and compile it again, but you will note that attributes which fall into this category I specifically added to the whitelist I wrote for this exact reason.

I agree that mixing user attributes and internal attributes is messy, but I would point out that since I am proposing we /not/ be allowed to parse internal attributes if they appear in rules, it would not do what you mention and interfere with backwards compatibility since the internal attributes would be entirely behind the abstraction boundary of the compiler. You would get the same issue, a parsing error, whether an attribute didn't exist or was internal, and adding a new attribute doesn't change the grammar of k used for parsing in any way.

As such, while I'm not opposed to maintaining a specific structured representation for internal attributes, I don't believe that it's that complicated or problematic to do what I originally mentioned.

Reply:
> As such, I don't really see the benefit of being able to parse internal attributes added by the compilation process. We do want to be able to parse attributes added by the /parsing/ process, since we probably want to be able to parse a definition, compile it, and then use the compiled and parsed definitions to modify the parsed definition and compile it again, but you will note that attributes which fall into this category I specifically added to the whitelist I wrote for this exact reason.

This makes sense to me. @radumereuta, does this cover what you were referring to?

It's not essential then, but it does seems like a structured `internal(_)` attribute could be a good way to handle the "we may wish to restrict their usage in user written code" attributes you listed in your other comment.

> I agree that mixing user attributes and internal attributes is messy, but I would point out that since I am proposing we /not/ be allowed to parse internal attributes if they appear in rules, it would not do what you mention and interfere with backwards compatibility since the internal attributes would be entirely behind the abstraction boundary of the compiler.

I agree that, once the whitelist is in effect, banning something like `[internal-att]` from appearing in the source code is completely fine. For all the user knows, `internal-att` is just not a known built-in, so there's no abstraction leak.

However, the issue is that we also would need to ban `[group(internal-att)]` because internally its represented the same as `[internal-att]`. This does actually break the abstraction because adding new internals could break backwards compatibility with existing user groups.

Reply:
Oh, I see. Then we should probably, in a follow-up change, put groups and nongroups in separate attribute pools.

Reply:
I think Dwight has a point about not allowing internal attributes through the parser.
It should work though through other means, like JSON ingest.